### PR TITLE
PUPIL-421, PUPIL-423: Enable worksheet downloads from the pupil intro

### PR DESCRIPTION
**Pull request by carlmw**
## Description

Music year: [1985](https://www.youtube.com/watch?v=GA4ROLQF4cs)

Enables the downloading of worksheets for new and legacy (🤞🏼) lessons

## Issue(s)
1. Though i've implemented this for both the new and legacy lessons there aren't any legacy lessons available to test with.
2. I've added a very basic error state should the download fail to start as a ([a simple alert](#failed-download)). If there are any examples of how to better handle this (or even a better message for the alert 😉 ) in the app elsewhere i'd be very happy to improve this.

## How to test

1. Go to https://deploy-preview-2237--oak-web-application.netlify.thenational.academy/pupils/programmes/combined-science-secondary-ks4-higher-aqa/units/measuring-waves/lessons/oscilloscope
3. Click on "Intro"
4. You should see the "Worksheet" card
5. Click "Download worksheet"
6. A zip file containing the worksheet should be downloaded 

## Screenshots

https://github.com/oaknational/Oak-Web-Application/assets/122096/e3641d8e-7a7c-49aa-8b08-584243ec107e

<span id="failed-download"></span>

## Failed download
https://github.com/oaknational/Oak-Web-Application/assets/122096/d1946e22-a0c5-4455-aad8-39e2dd5d8f96

## Checklist

- [ ] Added or updated tests where appropriate
- [ ] Manually tested across browsers / devices
- [ ] Considered impact on accessibility

### CODE REVIEW

File: src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx [hunk 1]
```diff
@@ -177,18 +184,35 @@ export const getStaticProps: GetStaticProps<
...
+          return getCaptionsFromFile(`${curriculumData.videoTitle}.vtt`);
+        }
+
+        return curriculumData.transcriptSentences;
+      })();
+
+      // Resolve the requests for the transcript and worksheet existence in parallel
+      const [transcriptSentences, downloadExistence] = await Promise.all([
+        resolveTranscriptSentences,
+        getDownloadResourcesExistence(
```
Comment (carlmw) [inline src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx:200]: Pinching this from the teacher's section — it makes a request to the downloads API (either the new or legacy one) to tell use if there is a worksheet for the lesson

File: src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx [hunk 2]
```diff
@@ -177,18 +184,35 @@ export const getStaticProps: GetStaticProps<
...
-      }
+      // For new content we need to fetch the captions file from gCloud and parse the result to generate
+      // the transcript sentences.
+      const resolveTranscriptSentences = (() => {
+        if (curriculumData.videoTitle && !curriculumData.isLegacy) {
+          return getCaptionsFromFile(`${curriculumData.videoTitle}.vtt`);
+        }
+
+        return curriculumData.transcriptSentences;
+      })();
```
Comment (carlmw) [inline src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx:195]: I originally had this as a big ternary but it was unreadable.

File: src/node-lib/curriculum-api-2023/queries/pupilLessonOverview/pupilLessonOverview.gql [hunk 3]
```diff
@@ -23,11 +23,10 @@ query pupilLessonOverview(
     supervisionLevel
     contentGuidance
     lessonEquipmentAndResources
-    worksheetUrl
     videoTitle
     videoMuxPlaybackId
     videoWithSignLanguageMuxPlaybackId
     transcriptSentences
-    lessonCohort
+    isLegacy
```
Comment (carlmw) [inline src/node-lib/curriculum-api-2023/queries/pupilLessonOverview/pupilLessonOverview.gql:30]: If i'm following the intention of this new field it is equivalent to `lesson.lessonCohort === LEGACY_COHORT` which means we pull this boolean out ourselves and use it directly.

File: src/node-lib/curriculum-api-2023/queries/pupilLessonOverview/pupilLessonOverview.gql [hunk 4]
```diff
@@ -23,11 +23,10 @@ query pupilLessonOverview(
     supervisionLevel
     contentGuidance
     lessonEquipmentAndResources
-    worksheetUrl
```
Comment (carlmw) [inline src/node-lib/curriculum-api-2023/queries/pupilLessonOverview/pupilLessonOverview.gql:26]: I've been told that `worksheetUrl` points to the original worksheet (Google Slides) which is only good for embedding.

File: src/components/PupilViews/PupilIntro/useWorksheetDownload.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+import downloadLessonResources from "@/components/SharedComponents/helpers/downloadAndShareHelpers/downloadLessonResources";
+import errorReporter from "@/common-lib/error-reporter";
+
+type WorksheetDownload = {
+  startDownload: () => Promise<void>;
+  isDownloading: boolean;
+};
+
+const reportError = errorReporter("useWorksheetDownload");
```
Comment (carlmw) [inline src/components/PupilViews/PupilIntro/useWorksheetDownload.ts:11]: I've seen this pattern throughout the codebase to report on runtime errors

**Comment by messmike** [comment]
@carlmw @rightsaidjames fyi, I've moved the other points I added earlier relating to the introduction page to this new ticket in the backlog: https://www.notion.so/oaknationalacademy/Introduction-page-updates-c7fba466fdba412c99eed87beeefe25c?pvs=4

**Review by rightsaidjames — APPROVED** [review]
Other than @messmike's feedback above, I think this is good to go from my perspective. We can revisit this when we are testing legacy lessons to ensure it works as expected and/or does the job.

**Comment by carlmw** [comment]
> I noticed the worksheet icon is the older one used in the card - it should use the one shown here (SVG is named 'Worksheet' in Figma): Screenshot 2024-02-12 at 11 22 18 am Screenshot 2024-02-12 at 11 23 49 am

@messmike i've updated the icon in https://github.com/oaknational/Oak-Web-Application/pull/2243

File: src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx (hunk 2 above)
Comment (kimon-satan) [reply src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx:195]: nice. I haven't seen self executing functions in ages but it's nice way of keeping variables as const without ternaries.

**Review by kimon-satan — APPROVED** [review]

File: src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx (hunk 2 above)
Comment (carlmw) [reply src/pages/pupils/programmes/[programmeSlug]/units/[unitSlug]/lessons/[lessonSlug]/index.tsx:195]: It's an affordance I miss from my days as a Rubyist

**Pull request merged**